### PR TITLE
[RemoveDIs] Add flag to control loading into new debug mode from bitcode

### PR DESCRIPTION
**Pull request by OCHyams**
--load-bitcode-into-experimental-debuginfo-iterators

      false: Convert to the old debug mode after reading.
      true: Upgrade to the new debug info format (*).
      unset: Same as false (for now).

(*) As of this patch it actually just means "don't convert to either mode after loading". Auto-upgrading will be implemented in an upcoming patch.

With this flag we can incrementally add support for RemoveDIs by overriding the "unset" behaviour in individual tools. The flag can be removed once all tools support the new debug info mode.

### CODE REVIEW

File: llvm/lib/Bitcode/Reader/BitcodeReader.cpp [hunk 1]
```diff
@@ -4276,9 +4283,11 @@ Error BitcodeReader::parseGlobalIndirectSymbolRecord(
...
                                  bool ShouldLazyLoadMetadata,
                                  ParserCallbacks Callbacks) {
-  // Force the debug-info mode into the old format for now.
-  // FIXME: Remove this once all tools support RemoveDIs.
-  TheModule->IsNewDbgInfoFormat = false;
+  // Load directly into RemoveDIs format if LoadBitcodeIntoNewDbgInforFormat
+  // has been set to true (default action: load into the old debug format).
+  TheModule->IsNewDbgInfoFormat =
+      UseNewDbgInfoFormat &&
+      LoadBitcodeIntoNewDbgInforFormat == cl::boolOrDefault::BOU_TRUE;
```
Comment (CarlosAlbertoEnciso) [inline llvm/lib/Bitcode/Reader/BitcodeReader.cpp:4290]: As you check `LoadBitcodeIntoNewDbgInforFormat` only for `BOU_TRUE`, may be just define it as `cl::opt<cl::bool>`

**Review by CarlosAlbertoEnciso — CHANGES_REQUESTED** [review]

**Comment by OCHyams** [comment]
Thanks @CarlosAlbertoEnciso - To answer your two inline questions about the `cl::opt` being a `boolOrDefault`: the default value is `BOU_UNSET`. So at the moment with the check `LoadBitcodeIntoNewDbgInforFormat == cl::boolOrDefault::BOU_TRUE` in BitcodeReader.cpp, `BOU_FALSE` and `BOU_UNSET` have the same meaning. However, we can have individual tools make this change:
```
if (LoadBitcodeIntoNewDbgInforFormat == cl::boolOrDefault::BOU_UNSET)
    LoadBitcodeIntoNewDbgInforFormat = cl::boolOrDefault::BOU_TRUE;
```
Which effectively means for that tool only, the default is changed from false to true.

It's useful to be able to give each tool control over this while we're implementing support, so it can be done incrementally. Eventually the flag will be removed.

**Comment by CarlosAlbertoEnciso** [comment]
> It's useful to be able to give each tool control over this while we're implementing support, so it can be done incrementally. Eventually the flag will be removed.

Thanks for the clarification on `cl::opt` being a `boolOrDefault`.

**Review by jryans — APPROVED** [review]

**Review by CarlosAlbertoEnciso — APPROVED** [review]
LGTM

**Comment by OCHyams** [comment]
Thanks both for the review

**Pull request merged**